File: Verify-Manual-octave/train/core/lrCostFunction.m
function [J, grad] = lrCostFunction(theta, X, y, lambda)
    m = length(y);

    J = 0;
    grad = zeros(size(theta));

    htheta = sigmoid(X*theta);
    J = -sum(y .* log(htheta) + (1 - y) .* log(1 - htheta))/m + lambda * sum(theta(2:end) .^ 2) / (2 * m);

    grad(1) = 1 / m * sum((htheta - y) .* X(:, 1));
    for i = 2:size(theta, 1)
        grad(i) = 1 / m * sum((htheta - y) .* X(:, i)) + lambda / m * theta(i);
    end
    grad = grad(:);
end

File: Verify-Manual-python/train/core/lrCostFunction.py
# coding: utf-8
from numpy import log, sum
from numpy import array, matrix

from sigmoid import sigmoid


def lrCostFunction(theta, *args):
    theta = matrix(theta).transpose()
    X, y, the_lambda = args
    m = y.shape[0]
    htheta = sigmoid(X*theta)
    J = -sum(array(y)*array(log(htheta)) + array((1-y))*array(log(1-htheta)))/m + the_lambda*sum(array(theta[1:]) * array(theta[1:]))/(2*m)
    return J
